.github/workflows/e2e_tests.yml

```
@@ -69,46 +90,144 @@ jobs:
           echo "ICP_PORT=$ICP_PORT" >> $GITHUB_OUTPUT
           echo "SWARM_HOST=$SWARM_HOST" >> $GITHUB_OUTPUT
 
-          # Update JSON file with jq
+          # Update config.json
           jq --arg swarmPort "$SWARM_PORT" \
-              --arg serverPort "$SERVER_PORT" \
-              --arg icpPort "$ICP_PORT" \
-              --arg swarmHost "$SWARM_HOST" \
-              '.network.swarmHost = ($swarmHost) |
-              .network.startSwarmPort = ($swarmPort | tonumber) |
-              .network.startServerPort = ($serverPort | tonumber) |
-              .protocolSandboxes[1].config.rpcUrl = "http://127.0.0.1:\($icpPort)"
-              ' e2e-tests/config/config.json > updated_config.json
+            --arg serverPort "$SERVER_PORT" \
+            --arg swarmHost "$SWARM_HOST" \
+            '.network.swarmHost = ($swarmHost) |
+            .network.startSwarmPort = ($swarmPort | tonumber) |
+            .network.startServerPort = ($serverPort | tonumber)' \
+            e2e-tests/config/config.json > updated_config.json
 
           mv updated_config.json e2e-tests/config/config.json
+      
+      - name: Build Contracts (if needed)
+        run: |
+          if [[ "${{ matrix.protocol }}" == "near" ]]; then
+            ./contracts/near/context-config/build.sh
+            ./contracts/near/context-proxy/build.sh
+          elif [[ "${{ matrix.protocol }}" == "stellar" ]]; then
+            ./contracts/stellar/context-config/build.sh
+            ./contracts/stellar/context-proxy/build.sh
+          fi
+      
+      - name: Install dfx
+        if: matrix.protocol == 'icp'
+        uses: dfinity/setup-dfx@main
+
+      - name: Cache candid-extractor
+        id: cache-candid
+        if: matrix.protocol == 'icp'
+        uses: actions/cache@v3
+        with:
+          path: ~/.cargo/bin/candid-extractor
+          key: ${{ runner.os }}-candid-extractor
+
+      - name: Install candid-extractor
+        id: install-candid
+        if: steps.cache-candid.outputs.cache-hit != 'true' && matrix.protocol == 'icp'
+        run: cargo install candid-extractor
 
-      - name: Deploy ICP local devnet
+      - name: Deploy Local Devnet (if needed)
         env:
-          ICP_PORT: ${{ steps.prepare_e2e_tests_config.outputs.ICP_PORT }}
+          ICP_PORT: ${{ steps.generate_ports.outputs.ICP_PORT }}
         run: |
-          echo "ICP_PORT=$ICP_PORT"
-          cargo install candid-extractor
-          cd ./contracts/icp/context-config
-          ./deploy_devnet.sh
-          cd ../../..
+          if [[ "${{ matrix.protocol }}" == "icp" ]]; then
+            echo "ICP_PORT=$ICP_PORT"
+            cd ./contracts/icp/context-config
+            ./deploy_devnet.sh
+            cd ../../..
```

Tracking `../../..` is easy to miss when things move around

```suggestion
            pushd ./contracts/icp/context-config
            ./deploy_devnet.sh
            popd
```